Refactor the edger8r

andschwa opened the pull request:
This refactoring pays back the technical debt of the edger8r having previously been a prototype. The main goal of the work were to make the edger8r easier to understand. To that end, most functions now read "forwards" instead of "backwards," that is, you can look at the list of strings and see skeleton of the code it is supposed to generate. Most formatting has been removed from the string content and instead applied to the lists of strings, again making it easier to parse the embedded C code.

All code was tested for current usage, with a fair amount of dead code identified and removed. As the refactoring progressed, better patterns were identified and implemented. Notably, embedded pattern matching to identify different pointer types was mostly replaced with maps onto lists instead filtered by predicates. Some code generation was identified as duplicate and so refactored into reusable functions; conversely some code which had no business being independent functions was "unfactored" to read more plainly.

The generated files now (mostly) pass `format-code`, with the exception of whitespace on empty lines and extra-long lines, both problems not being worth fixing for generated code. Also of note, when there is nothing to emit, a comment to that effect is now produced instead of an extra blank line, making it obvious to the reader that indeed no code should exist there. All generated comments and headers are now consistent and easy to identify, with more section headers introduced for easier reading of the edge routines.

Due to the nature of the work, this code is best reviewed by commit instead of by the whole diff. Most commits in the series should build without further commits, except where noted.

Still TODO: I identified a bad pattern I'd introduced in the predicates where I'm matching `(ptype, decl)` and throwing away the `decl`, as these functions became increasingly reused it was made obvious their signature should change to just `ptype`, but it was too late to do this retroactively. Expect another commit or two on Monday, with that change, the addition of a unit test for errno propagation, a cleanup of the member decl generation logic (note internal TODOs), and maybe a commit to group functions logically (now that it is possible to do so).

andschwa commented:
This will make implementing deep-copy (and other edger8r changes) _worlds_ easier (for me).

andschwa commented:
If you care to, take a look at the generated code of e.g. the edger8r tests before and after this chain. It's so much better.

anakrish commented:
Snippet from build/tests/echo/enc/echo_args.h
```c
/* No errno support. */

#include <openenclave/bits/result.h>

/**** User includes. ****/
/* There were no user includes. */

/**** User defined types in EDL. ****/
/* There were no user defined types. */
/**** ECALL marshalling structs. ****/
``` 

A newline before ECALL marshalling structs would make this more readable.
Not clear what 'No errno support' means.

anakrish commented:
```c
/**** Trusted function IDs ****/
enum
{
    fcn_id_enc_echo = 0,
    fcn_id_trusted_call_id_max = OE_ENUM_MAX
};

/**** Untrusted function IDs. ****/
enum
{
    fcn_id_host_echo = 0,
    fcn_id_untrusted_call_max = OE_ENUM_MAX
};
```

Does refactoring make it easy to prefix the enums with the edl file name? Making enums globally unique allows the host to include multiple edl _u.h files easily.

andschwa commented:
> A newline before ECALL marshalling structs would make this more readable.

Ah, thanks! Must have missed that.

> Not clear what 'No errno support' means.

As in, `propogate_errno` was not true. I can make this comment more explicit.

anakrish commented:
>> The generated files now (mostly) pass format-code

Though not critical, I can see some code that is not format-code compatible.
Inconsistent spacing, comments longer than 80 chars etc.
```c
 host_echo_args_t _args, *_pargs_in = NULL, *_pargs_out=NULL;
```

andschwa commented:
> ```c
> ```c
>  host_echo_args_t _args, *_pargs_in = NULL, *_pargs_out=NULL;
> ```
> 
> 
>       
> ```

Are you sure you generated? I specifically remember fixing `*_pargs_out=NULL`.

Also, I don't believe line length limits are worth fixing:

> with the exception of whitespace on empty lines and extra-long lines

The other notable exception is `if (foo) blah` on one line, where one line is better even though it's too long, because it's repetitive, generated code.

anakrish commented:
Yes you are right RE _pargs_out. I had gotten my master/PR tabs mixed up.

anakrish commented:
There seems to be some indentation issue in some cases:

```c
    /* Compute input buffer size. Include in and in-out parameters. */
    OE_ADD_SIZE(_input_buffer_size, sizeof(host_echo_args_t));
    if (in) OE_ADD_SIZE(_input_buffer_size, _args.in_len * sizeof(char));
if (str1) OE_ADD_SIZE(_input_buffer_size, _args.str1_len * sizeof(char));
if (str3) OE_ADD_SIZE(_input_buffer_size, sizeof(char[100]));
```

andschwa commented:
> ```c
> /**** Trusted function IDs ****/
> enum
> {
>     fcn_id_enc_echo = 0,
>     fcn_id_trusted_call_id_max = OE_ENUM_MAX
> };
> 
> /**** Untrusted function IDs. ****/
> enum
> {
>     fcn_id_host_echo = 0,
>     fcn_id_untrusted_call_max = OE_ENUM_MAX
> };
> ```
> 
> Does refactoring make it easy to prefix the enums with the edl file name? Making enums globally unique allows the host to include multiple edl _u.h files easily.

Do you mean something like (for `foo.edl`):

```
enum
{
    foo_fcn_id_host_echo = 0,
    foo_fcn_id_untrusted_call_max = OE_ENUM_MAX
};
```

Or does the `enum` need a name too?

anakrish commented:
> 
> Do you mean something like (for `foo.edl`):
> 
> ```
> enum
> {
>     foo_fcn_id_host_echo = 0,
>     foo_fcn_id_untrusted_call_max = OE_ENUM_MAX
> };
> ```

Yes. This would be ideal.

anakrish commented:
We could space the following a bit better
```c
    /* Unmarshal return value and out, in-out parameters. */
    *_retval = _pargs_out->_retval;
    OE_READ_OUT_PARAM(out, (size_t)(sizeof(char[100])));
    /* Propagate errno. */
    /* Errno not enabled. */
    _result = OE_OK;
```

andschwa commented:
> ```c
> OE_ADD_SIZE
> ```

Hunting for this...have a feeling I broke it in my last commit.

andschwa commented:
> There seems to be some indentation issue in some cases:
> 
> ```c
>     /* Compute input buffer size. Include in and in-out parameters. */
>     OE_ADD_SIZE(_input_buffer_size, sizeof(host_echo_args_t));
>     if (in) OE_ADD_SIZE(_input_buffer_size, _args.in_len * sizeof(char));
> if (str1) OE_ADD_SIZE(_input_buffer_size, _args.str1_len * sizeof(char));
> if (str3) OE_ADD_SIZE(_input_buffer_size, sizeof(char[100]));
> ```

Fixed. Thanks for doing all that scrolling. This broke when I refactored this into `oe_compute_input_buffer_size`; since now I'm returning a single string back to a list that has its indentation applied when it's concatted. Still looking for a better way to handle that, but it's fixed with a note now.

andschwa commented:
> We could space the following a bit better
> 
> ```c
>     /* Unmarshal return value and out, in-out parameters. */
>     *_retval = _pargs_out->_retval;
>     OE_READ_OUT_PARAM(out, (size_t)(sizeof(char[100])));
>     /* Propagate errno. */
>     /* Errno not enabled. */
>     _result = OE_OK;
> ```

Like:

```c
    /* Unmarshal return value and out, in-out parameters. */
    *_retval = _pargs_out->_retval;
    OE_READ_OUT_PARAM(out, (size_t)(sizeof(char[100])));

    /* Propagate errno. */
    /* Errno not enabled. */

    _result = OE_OK;
```
?

anakrish commented:
> Like:
> 
> ```c
>     /* Unmarshal return value and out, in-out parameters. */
>     *_retval = _pargs_out->_retval;
>     OE_READ_OUT_PARAM(out, (size_t)(sizeof(char[100])));
> 
>     /* Propagate errno. */
>     /* Errno not enabled. */
> 
>     _result = OE_OK;
> ```
> 
> ?
Looks good.

anakrish commented:
I noticed that the generated files no longer have a terminating newline character.

ISO C requires a newline character at the end of the file. 
If you put "int foo() { return 1; }" in a file foo.c and compile it using clang -c -Wpedantic foo.c, it will emit a warning:
```bash
$ clang -c   -Wpedantic foo.c
foo.c:4:2: warning: no newline at end of file [-Wnewline-eof]
}
 ^
1 warning generated.
```

Any reason why we want to make the generated code non ISO C compatible?
@CodeMonkeyLeet If I remember correctly, at one point in time we had an automatic check to prevent check-in of non newline-terminated files.

andschwa commented:
> I noticed that the generated files no longer have a terminating newline character.
> 
> ISO C requires a newline character at the end of the file.
> If you put "int foo() { return 1; }" in a file foo.c and compile it using clang -c -Wpedantic foo.c, it will emit a warning:
> 
> ```shell
> $ clang -c   -Wpedantic foo.c
> foo.c:4:2: warning: no newline at end of file [-Wnewline-eof]
> }
>  ^
> 1 warning generated.
> ```
> 
> Any reason why we want to make the generated code non ISO C compatible?
> @CodeMonkeyLeet If I remember correctly, at one point in time we had an automatic check to prevent check-in of non newline-terminated files.

Oh I forgot about that. Yes, I need to make sure they all end with newlines. Thanks!

andschwa commented:
> > Do you mean something like (for `foo.edl`):
> > ```
> > enum
> > {
> >     foo_fcn_id_host_echo = 0,
> >     foo_fcn_id_untrusted_call_max = OE_ENUM_MAX
> > };
> > ```
> 
> Yes. This would be ideal.

Yeah, I'd say pretty easy considering I have a commit implementing this and passing tests right now. That said, I'd like to clean it up to pass through a curried version of get_function_id instead of passing through the enclave name itself, seems like it'd be cleaner.

andschwa commented:
> I noticed that the generated files no longer have a terminating newline character.
> 
> ISO C requires a newline character at the end of the file.
> If you put "int foo() { return 1; }" in a file foo.c and compile it using clang -c -Wpedantic foo.c, it will emit a warning:
> 
> ```shell
> $ clang -c   -Wpedantic foo.c
> foo.c:4:2: warning: no newline at end of file [-Wnewline-eof]
> }
>  ^
> 1 warning generated.
> ```
> 
> Any reason why we want to make the generated code non ISO C compatible?
> @CodeMonkeyLeet If I remember correctly, at one point in time we had an automatic check to prevent check-in of non newline-terminated files.

Fixed, thanks for the catch. It was the C files I'd forgotten to newline-terminate.

andschwa commented:
> We could space the following a bit better
> 
> ```c
>     /* Unmarshal return value and out, in-out parameters. */
>     *_retval = _pargs_out->_retval;
>     OE_READ_OUT_PARAM(out, (size_t)(sizeof(char[100])));
>     /* Propagate errno. */
>     /* Errno not enabled. */
>     _result = OE_OK;
> ```

Fixed.

andschwa commented:
> A newline before ECALL marshalling structs would make this more readable.

Fixed (I actually have a TODO to make all these newlines a bit cleaner here, but worked around it for now.)

> Not clear what 'No errno support' means.

Fixed. I clarified all the messages; let me know what you think.

andschwa commented:
@gupta-ak I don't know how you did any work in the edger8r before this 😅

anakrish commented:
Couple more comments re generated code. After this I am going to look at the Ocaml implementation:

1. The include guard for _args.h needs to have the EDGER8R_ prefix like _t.h and _u.h.
2. Errono propagation not enabled makes sense on a per function basis and adds value when reading each function's generated implementation. However, it feels out of place in _args.h

anakrish commented on tools/oeedger8r/Emitter.ml:
```diff
@@ -217,102 +231,143 @@ let oe_get_param_size (ptype, decl, argstruct) =
-  | UnionDef u -> emit_struct_or_union os u true
-  | EnumDef e -> emit_enum os e
+let emit_composite_type =
+  let emit_struct_or_union (s : struct_def) (union : bool) =
```
I like the way this looks.

andschwa replied:
Me too, minus the use of bool for union/struct...but apparently types aren't that strongly represented in OCaml. There's a lot of code I found could be simplified if I could simply pass a type constructor _as_ an argument.

anakrish commented on tools/oeedger8r/Emitter.ml:
```diff
@@ -217,102 +231,143 @@ let oe_get_param_size (ptype, decl, argstruct) =
+               ( match value with
+               | EnumVal (AString s) -> " = " ^ s
+               | EnumVal (ANumber n) -> " = " ^ string_of_int n
+               | EnumValNone -> "" ) )
```
Can you take a follow up item to see if we need to add a final enumeration member whose value equals to OE_ENUM_MAX?

andschwa replied:
Filed #1767.

anakrish commented on tools/oeedger8r/Emitter.ml:
```diff
@@ -325,153 +380,139 @@ let oe_gen_args_header (ec : enclave_content) (dir : string) =
+  if params <> [] then String.concat "\n    " params
+  else "/* There were no out nor in-out parameters. */"
+
+let oe_serialize_buffer_inputs (plist : pdecl list) =
```
It feels like these duplication in these 3 functions can be avoided.
In future.

andschwa replied:
I know, I felt that way while I was writing them, but just enough is different that it makes it difficult to deduplicate further. Their filter predicates are different (so their comments are different) and what they're printing is different. I opted for readability, but certainly it can be improved.

anakrish commented on tools/oeedger8r/Emitter.ml:
```diff
@@ -325,153 +380,139 @@ let oe_gen_args_header (ec : enclave_content) (dir : string) =
+      ; "if ((_result = _pargs_out->_result) != OE_OK)"
+      ; "    goto done;"
+      ; ""
+      ; "/* Currently exactly _output_buffer_size bytes must be written. */"
```
In the future we can tweaks this to support variable length outputs.

andschwa replied:
That'd be cool.

anakrish commented on tools/oeedger8r/Emitter.ml:
```diff
@@ -483,435 +524,371 @@ let get_cast_from_mem_expr (ptype, decl) =
+      [ "/* Set out and in-out pointers. */"
+      ; "/* In-out parameters are copied to output buffer. */"
+      ; ( if params <> [] then String.concat "\n    " params
+        else "/* There were no out nor in-out parameters. */" ) ]
```
Can you double check the grammar here? 
We were taught that it was "neither..nor"

andschwa replied:
Yeah it [seems](https://www.wikihow.com/Use-nor) to [be](https://thewritepractice.com/how-to-use-either-neither-or-and-nor-correctly/) correct. It's not as common, but you can use 'nor' without 'neither' if it's paired with another negative (in this case, 'no').

anakrish commented on tools/oeedger8r/Emitter.ml:
```diff
@@ -483,435 +524,371 @@ let get_cast_from_mem_expr (ptype, decl) =
+            fd.plist)
+     in
+     if params <> [] then String.concat "\n" params
+     else "    /* There were no in nor in-out string parameters. */")
```
Double-check 'nor' usage.

andschwa commented on tools/oeedger8r/Emitter.ml:
```diff
@@ -110,13 +101,34 @@ let mk_ms_member_decl (pt : parameter_type) (declr : declarator)
+  | PTPtr (_, a) -> a.pa_chkptr && a.pa_direction = PtrInOut
+
+let is_str_ptr (ptype, _) =
+  match ptype with PTVal _ -> false | PTPtr (_, a) -> a.pa_isstr
```
I'm just using `ocamlformat` with the defaults; we can adjust it if you really want, but it seemed easier to take the decision out of our hands 😄

andschwa commented on tools/oeedger8r/Emitter.ml:
```diff
@@ -149,21 +162,22 @@ let oe_gen_marshal_struct_impl (fd : func_decl) (errno : string)
 
 let oe_gen_ecall_marshal_struct (tf : trusted_func) =
-  oe_gen_marshal_struct_impl tf.tf_fdecl "" true
+  oe_gen_marshal_struct_impl tf.tf_fdecl "" true ^ "\n"
```
`ocamlformat`...

andschwa replied:
(Literally just tried putting parens around it, on save my hook that runs `ocamlformat` removed them 😅)

andschwa commented on tools/oeedger8r/Emitter.ml:
```diff
@@ -217,102 +231,143 @@ let oe_get_param_size (ptype, decl, argstruct) =
-  let params_str =
-    if List.length fd.plist = 0 then "void" else get_plist_str fd
+  let get_plist_str (plist : pdecl list) =
+    if List.length plist = 0 then "void"
```
You know, I noticed some funny stuff with checking `Void`, I'd punted it for a later fix.

andschwa replied:
Oh, no, this is not equivalent to `fd.rtype = Void`, as it is not the return type. This is checking the function takes zero arguments (plist), and if so, generates the correct *C* (not C++) code to say `foo(void)` instead of `foo()`.

andschwa commented:
> * The include guard for _args.h needs to have the EDGER8R_ prefix like _t.h and _u.h.

I checked master and this is an existing bug, fixed now though.

andschwa commented:
> 2\. Errono propagation not enabled makes sense on a per function basis and adds value when reading each function's generated implementation. However, it feels out of place in _args.h

It's in place of `#include <errno.h>`. Let me include that string in the comment itself to make it more obvious.

andschwa commented:
@anakrish Do you want to include #1765 in this? I didn't get a chance to add an explicit test for it, but it is also doesn't break anything so we certainly could chunk it since this is under development anyway.

andschwa commented:
bors try

anakrish commented:
> @anakrish Do you want to include #1765 in this? I didn't get a chance to add an explicit test for it, but it is also doesn't break anything so we certainly could chunk it since this is under development anyway.

Let us hold off on that one.  This PR is a net positive. Bringing in #1765 without a test feels like we are taking away from this PR. There is no immediate urgency for #1765

anakrish commented:
> 2. Errono propagation not enabled makes sense on a per function basis and adds value when reading each function's generated implementation. However, it feels out of place in _args.h
> 
> It's in place of #include <errno.h>. Let me include that string in the comment itself to make it more obvious.

Ah ok. It never occurred to me that the message was taking the place of the errno header.

andschwa commented:
Noting that my TODOs in the PR description are future TODOs and don't block this, merging!

bors r+

Thanks for the reviews Anand!
